wrapper.go

```diff
@@ -0,0 +1,135 @@
+}
+
+func getDefaultDimensions(ctx context.Context) (map[string]string, error) {
+	if lambdaContext, ok := lambdacontext.FromContext(ctx); ok {
+		if strings.TrimSpace(lambdaContext.InvokedFunctionArn) == "" {
+			return nil, fmt.Errorf("lambda function arn cannot be blank")
+		}
+		// Expected function arn format arn:aws:lambda:us-east-1:accountId:function:functionName:$LATEST
+		arnTokens := strings.Split(lambdaContext.InvokedFunctionArn, ":")
+		dimensions := map[string]string{
+			"aws_function_version": lambdacontext.FunctionVersion,
+			"aws_function_name":    lambdacontext.FunctionName,
+			"aws_region":           arnTokens[3],
+			"aws_account_id":       arnTokens[4],
+			"metric_source":        "lambda_wrapper",
+			//'function_wrapper_version': name + '_' + version,
+		}
+		if os.Getenv("AWS_EXECUTION_ENV") != "" {
+			dimensions["ws_execution_env"] = os.Getenv("AWS_EXECUTION_ENV")
+		}
+		if arnTokens[5] == "function" {
+			if len(arnTokens) == 8 {
+				dimensions["aws_function_qualifier"] = arnTokens[7]
+				arnTokens[7] = lambdacontext.FunctionVersion
+			} else if len(arnTokens) == 7 {
+				arnTokens = append(arnTokens, lambdacontext.FunctionVersion)
+			}
+			dimensions["lambda_arn"] = strings.Join(arnTokens, ":")
+		} else if arnTokens[5] == "event-source-mappings" {
+			dimensions["event_source_mappings"] = arnTokens[6]
+			dimensions["lambda_arn"] = lambdaContext.InvokedFunctionArn
+		}
+		return dimensions, nil
+	}
+	return nil, fmt.Errorf("failed to get *LambdaContext from %+v", ctx)
+}
+
+func getNonErrorReturnType(handlerFunc interface{}) reflect.Type {
+	handlerFuncType := reflect.TypeOf(handlerFunc)
+	if handlerFuncType.NumOut() == 2 {
+		return handlerFuncType.Out(0)
+	}
+	return nil
+}
+
+func (hfw *handlerFuncWrapper) sendInvocationsDatapoint() {
+	dp := datapoint.Datapoint{Metric: "function.invocations", Value: datapoint.NewIntValue(1), MetricType: datapoint.Counter}
+	dp.Dimensions = datapoint.AddMaps(hfw.defaultDimensions, dp.Dimensions)
+	sendDatapoint(nil, &dp)
+}
+
+func (hfw *handlerFuncWrapper) sendColdStartsDatapoint() {
+	if !hfw.notColdStart {
+		dp := datapoint.Datapoint{Metric: "function.cold_starts", Value: datapoint.NewIntValue(1), MetricType: datapoint.Counter}
+		dp.Dimensions = datapoint.AddMaps(hfw.defaultDimensions, dp.Dimensions)
+		sendDatapoint(nil, &dp)
+	}
+	hfw.notColdStart = true
+}
+
+func (hfw *handlerFuncWrapper) sendDurationDatapoint(elapsed time.Duration) {
+	dp := datapoint.Datapoint{Metric: "function.duration", Value: datapoint.NewFloatValue(elapsed.Seconds()), MetricType: datapoint.Gauge}
+	dp.Dimensions = datapoint.AddMaps(hfw.defaultDimensions, dp.Dimensions)
+	sendDatapoint(nil, &dp)
+}
+
+func (hfw *handlerFuncWrapper) sendErrorsDatapoint() {
+	dp := datapoint.Datapoint{Metric: "function.errors", Value: datapoint.NewIntValue(1), MetricType: datapoint.Counter}
+	dp.Dimensions = datapoint.AddMaps(hfw.defaultDimensions, dp.Dimensions)
+	sendDatapoint(nil, &dp)
```

Comment:
So there are up to four separate datapoint requests going out, one per datapoint.  I would batch them all together into a single request to minimize network traffic and computation required.